doc: board_porting: Move the board extensions section

57300 opened the pull request:
It was awkwardly placed in the middle of the "Write your devicetree" section, breaking up the flow of the guide. A more natural placement would be underneath "Contributing your board", considering that this feature deals with boards already submitted to Zephyr.

Let's relocate this part and swap two of its paragraphs in the process, to motivate the feature before describing it. Tweak a few sentences and add a basic example, to fit in with the overall document a little better.

57300 committed:
doc: board_porting: Extend the board extensions section

Tweak a few sentences and add a basic example, to fit in with the
overall document a little better. Also, fix an invalid cross-reference.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
doc: board_porting: Move the board extensions section

It was awkwardly placed in the middle of the "Write your devicetree"
section, breaking up the flow of the guide. A more natural placement
would be underneath "Contributing your board", considering that this
feature deals with boards already submitted to Zephyr.

Let's relocate this part and swap two of its paragraphs in the process,
to motivate the feature before describing it.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

kartben commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -810,3 +779,44 @@ There are some extra things you'll need to do:
+
+Boards already supported by Zephyr can be extended by downstream users, such as
+``example-application`` or vendor SDKs. In some situations, certain hardware
+description or `choices <devicetree-chosen-nodes>`_ can not be added in the
```
```suggestion
description or :ref:`choices <devicetree-chosen-nodes>` can not be added in the
```

kartben commented on doc/hardware/porting/board_porting.rst:
```diff
@@ -810,3 +779,44 @@ There are some extra things you'll need to do:
+   └── plank_<revision>.overlay  # optional
+
+A board extension directory must follow the naming structure of the original
+board to extend. It may contain Kconfig fragments and/or devicetree overlays.
```
nit/style (feel free to ignore)
```suggestion
board it extends. It may contain Kconfig fragments and/or devicetree overlays.
```

kartben reviewed (changes requested):
Great, thanks! Just spotted an (existing) error in a reference.

Two additional notes:
- The structure of this document is still far from great, and super "flat", and it can probably use a more significant overhaul in the future (see top level headings below [^1])
- Bonus points if you use the opportunity of fixing the dead link mentioned in my review comment to update `boards/index.rst` and have it cross-reference the porting guide, as it's something that I think is cruelly missing.

Thanks!

[^1]: top-level headings.
    > * Boards, SoCs, etc.
    > * Make sure your SoC is supported
    > * Create your board directory
    > * Write your devicetree
    > * Write Kconfig files
    > * Build, test, and fix
    > * General recommendations
    > * Flash and debug support
    > * Multiple board revisions
    > * Custom revision.cmake files
    > * Contributing your board
    > * Board extensions

57300 commented:
Thanks for the suggestions @kartben!

> * The structure of this document is still far from great, and super "flat", and it can probably use a more significant overhaul in the future

I personally agree. I just jumped to fix one thing that stuck out to me while skimming the page. I think I'll pass on making further improvements for now, because I'm not a very efficient writer.

> * Bonus points if you use the opportunity (...) to update `boards/index.rst` and have it cross-reference the porting guide, as it's something that I think is cruelly missing.

Sorry, I'm not sure what you mean. Can you clarify the exact change you're suggesting for `boards/index.rst`?

kartben commented:
> I personally agree. I just jumped to fix one thing that stuck out to me while skimming the page. I think I'll pass on making further improvements for now, because I'm not a very efficient writer.

I should have made it clear that I was in no way asking you to do any further improvements :)

> > * Bonus points if you use the opportunity (...) to update `boards/index.rst` and have it cross-reference the porting guide, as it's something that I think is cruelly missing.
> 
> Sorry, I'm not sure what you mean. Can you clarify the exact change you're suggesting for `boards/index.rst`?

Right now it says

> To add support documentation for a new board, please use the template available
under :zephyr_file:`doc/templates/board.tmpl`

But I think it'd be nice to add a reference to the porting guide and rework it along the lines of:

> To add support for a new board, follow the :ref:`board_porting_guide`. If you are contributing the board to Zephyr, please use the template available under :zephyr_file:`doc/templates/board.tmpl` to properly add support documentation for it.

57300 committed:
doc: boards: Link to the board porting guide

This is the more useful document to point newer Zephyr users to.

The support documentation template is already mentioned in the guide
(under "Contributing your board"), but leave the existing link to it for
quick reference, just changing a few words.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
